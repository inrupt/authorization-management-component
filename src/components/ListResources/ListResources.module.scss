@use "../../../scss/_variables" as *;

@import "../../../node_modules/bootstrap/scss/mixins";

.resource-div {
  width: 100%;
  background-color: transparent;
  color: $primary-black;
  display: flex;
}

.adjacent-icon {
  width: 27.5;
  padding-right: 5px;
}

.name-content {
  background-color: transparent;
  width: fit-content;
}

.chevron-icon {
  padding-left: 5px;
  cursor: pointer;
  width: 5px;
}

.loader {
  margin: 5px;
  cursor: pointer;
  width: 15px;
  height: 15px;
}

.resource-container {
  width: 100%;
  outline: 1px solid $secondary-border-gray;
  border-radius: 5px;
  padding: 10px;
}
