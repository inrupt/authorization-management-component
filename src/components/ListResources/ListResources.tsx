//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
import { useEffect, useState } from "react";
import styles from "./ListResources.module.scss";
import FileIcon from "../FileIcon/FileIcon";

export interface Props {
  key: string;
  type: "Resource" | "Container";
  name: string;
  contains: () => Promise<Props[]>;
  depth?: number;
  icon?: boolean;
  resourceIri: string;
}

export default function ListResources({
  depth: tempDepth = 0,
  type,
  name,
  icon = true,
  resourceIri,
  contains,
}: Props) {
  const [expanded, setExpanded] = useState(false);
  const [expandAttempted, setExpandAttempted] = useState(false);
  const [containedResources, setContainedResources] = useState<false | Props[]>(
    false
  );
  const depth = tempDepth ?? 0;

  useEffect(() => {
    if (expandAttempted) {
      contains()
        .then((res) => {
          setContainedResources(res);
        })
        .catch(() => {
          setExpanded(false);
          setExpandAttempted(false);
        });
    }
  }, [expandAttempted, contains]);

  const onClick = () => {
    setExpanded(!expanded);
    if (!expanded && !expandAttempted) setExpandAttempted(true);
  };

  return (
    <>
      <div className={styles["resource-div"]}>
        <div
          className={styles["adjacent-icon"]}
          style={{
            marginLeft: icon ? 25 * depth : 0,
            paddingRight: icon ? 6 : 0,
          }}
        >
          {icon && <FileIcon expanded={expanded} resourceIri={resourceIri} />}
        </div>
        <div className={styles["name-content"]}>{name}</div>
        {type === "Container" &&
          (expanded && !containedResources ? (
            <div
              className={`spinner-border ${styles.loader}`}
              role="button"
              onClick={onClick}
              onKeyDown={onClick}
              tabIndex={0}
            >
              <span className="visually-hidden">Loading...</span>
            </div>
          ) : (
            <i
              className={`bi bi-chevron-${expanded ? "down" : "right"} ${
                styles["chevron-icon"]
              }`}
              onClick={(e) => {
                e.stopPropagation();
                onClick();
              }}
              onKeyDown={onClick}
              role="button"
              tabIndex={0}
              title={expanded ? "collapse" : "expand"}
            />
          ))}
      </div>

      {expanded &&
        containedResources &&
        containedResources.map((prop) => {
          return (
            <ListResources
              {...prop}
              depth={icon ? depth + 1 : depth}
              key={prop.key}
            />
          );
        })}
    </>
  );
}
