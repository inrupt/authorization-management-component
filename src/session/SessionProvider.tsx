// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import type { ReactElement, ReactNode } from "react";
import {
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import type { Session } from "@inrupt/solid-client-authn-browser";
import {
  fetch,
  getDefaultSession,
  login,
  logout,
} from "@inrupt/solid-client-authn-browser";

import { useRouter } from "next/router";
import { getPodUrlAllFrom, getProfileAll } from "@inrupt/solid-client";
import { getAccessApiEndpoint } from "@inrupt/solid-client-access-grants";
import { getVerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc";
import type { VerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc/dist/common/common";
import {
  WebIdNotAvailable,
  DiscoveryNotAvailable,
  AccessGrantsNotSupported,
} from "../../src/errors/customErrors";
import { ProfileModal } from "../components/ProfileModal/ProfileModal";
import { ErrorModal } from "../components/ErrorModal/ErrorModal";

type AuthState =
  | "waiting"
  | "authenticating"
  | "unauthenticated"
  | "authenticated"
  | "pre-authenticating";

export interface ISessionContext {
  login: typeof login;
  logout: typeof logout;
  session: Session;
  fetch: typeof fetch;
  sessionRequestInProgress: boolean;
  state: AuthState;
  accessEndpoint?: string;
  endpointConfiguration?: VerifiableCredentialApiConfiguration;
}

export const createContextBase: () => ISessionContext = () => ({
  login,
  logout,
  fetch,
  session: getDefaultSession(),
  sessionRequestInProgress: true,
  accessEndpoint: undefined,
  state: "waiting",
  endpointConfiguration: undefined,
});

export const SessionContext =
  createContext<ISessionContext>(createContextBase());

/* eslint react/require-default-props: 0 */
export interface ISessionProvider {
  children: ReactNode;
  onError?: (_error: Error) => void;
  restorePreviousSession?: boolean;
}

/**
 * @param session
 * @param callback The callback should mutate react state to trigger re-rendering.
 */
const discoverAccessEndpoint = async (
  session: Session,
  callback: (endpoint: string) => void,
): Promise<void> => {
  if (typeof session.info.webId !== "string") {
    return;
  }
  let profiles;
  try {
    profiles = await getProfileAll(session.info.webId, {
      fetch: session.fetch,
    });
  } catch (e: unknown) {
    throw new WebIdNotAvailable(
      `An error occured when looking up WebID ${session.info.webId}: ${e}`,
    );
  }
  const pods = getPodUrlAllFrom(profiles, session.info.webId);
  if (pods.length < 1) {
    throw new Error(
      `No Pods could be discovered from WebID ${session.info.webId}`,
    );
  }
  // FIXME arbitrarily pick the first Pod.
  const [podRoot] = pods;
  let endpoint;
  try {
    endpoint = await getAccessApiEndpoint(podRoot);
  } catch (e: unknown) {
    throw new DiscoveryNotAvailable(
      `Access endpoint discovery for Pod ${podRoot} failed: ${e}`,
    );
  }
  callback(endpoint);
};

const HintModal = ({
  error,
  modalStatus,
  modalControl,
}: {
  error?: Error;
  modalStatus: boolean;
  modalControl: (status: boolean) => void;
}) => {
  if (error === undefined) {
    return <></>;
  }
  if (error instanceof WebIdNotAvailable) {
    return (
      <ProfileModal
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  if (error instanceof DiscoveryNotAvailable) {
    return (
      <ErrorModal
        err={
          new AccessGrantsNotSupported(
            "Your Pod Provider doesn't appear to support Access Grants. Please retry using a compliant Pod Provider.",
            {
              cause: error,
            },
          )
        }
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  return (
    <ErrorModal
      err={error}
      isOpen={modalStatus}
      onClose={() => {
        modalControl(false);
      }}
    />
  );
};

/**
 * Used to provide session data to child components through context.
 */
export default function SessionProvider({
  children,
  onError,
  restorePreviousSession,
}: ISessionProvider): ReactElement {
  const { replace, isReady } = useRouter();
  const session = useMemo(getDefaultSession, []);

  const [state, setState] = useState<AuthState>(
    session.info.isLoggedIn ? "authenticated" : "waiting",
  );

  const [accessEndpoint, setAccessEndpoint] = useState<string>();
  const [accessEndpointConfig, setAccessEndpointConfig] =
    useState<VerifiableCredentialApiConfiguration>();
  const [modalOpen, setModalOpen] = useState<boolean>(true);
  const [errorState, setErrorState] = useState<Error>();

  const isAuthenticated = state === "authenticated";
  const isUnAuthenticated = state === "unauthenticated";

  useEffect(() => {
    if (isReady) {
      if (state === "waiting") {
        setState("authenticating");
      } else if (state === "authenticating") {
        session
          .handleIncomingRedirect({
            url: window.location.href,
            restorePreviousSession,
          })
          .then((res) => {
            setState(res?.isLoggedIn ? "authenticated" : "unauthenticated");
          })
          .catch((error) => {
            if (onError) {
              onError(error as Error);
            } else {
              throw error;
            }
            setState(
              session.info?.isLoggedIn ? "authenticated" : "unauthenticated",
            );
          });
      }
    }
  }, [state, isReady, onError, replace, session, restorePreviousSession]);

  useEffect(() => {
    session.events.on("logout", () => {
      if (isAuthenticated || isUnAuthenticated) {
        setState("waiting");
      }
    });
  }, []);

  useEffect(() => {
    if (state === "authenticated") {
      discoverAccessEndpoint(session, (endpoint: string) => {
        setAccessEndpoint(endpoint);
      }).catch((e) => {
        if (e instanceof Error) {
          setErrorState(e);
        } else {
          console.error(`An error occured: ${e}`);
        }
      });
    }
  }, [state]);

  useEffect(() => {
    if (typeof accessEndpoint === "string") {
      getVerifiableCredentialApiConfiguration(accessEndpoint).then(
        setAccessEndpointConfig,
        (error: unknown) => {
          if (error instanceof Error) {
            setErrorState(error);
          } else {
            console.error(`An error occured: ${error}`);
          }
        },
      );
    }
  }, [accessEndpoint]);

  const loginCallback = useCallback(
    async (options: Parameters<Session["login"]>[0]) => {
      if (isUnAuthenticated) {
        setState("pre-authenticating");
        try {
          await session.login(options);
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        } finally {
          setState("unauthenticated");
        }
      }
    },
    [isUnAuthenticated, onError, session],
  );

  const logoutCallback = useCallback(
    async (options: Parameters<Session["logout"]>[0]) => {
      if (isAuthenticated) {
        try {
          await session.logout(options);
          setState("unauthenticated");
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        }
      }
    },
    [isAuthenticated, onError, session],
  );

  const value = useMemo<ISessionContext>(
    () => ({
      login: loginCallback,
      logout: logoutCallback,
      fetch: session.fetch,
      session,
      sessionRequestInProgress:
        state === "waiting" || state === "authenticating",
      state,
      accessEndpoint,
      endpointConfiguration: accessEndpointConfig,
    }),
    [
      loginCallback,
      logoutCallback,
      state,
      session,
      accessEndpoint,
      accessEndpointConfig,
    ],
  );

  return (
    <SessionContext.Provider value={value}>
      <>
        <HintModal
          modalControl={setModalOpen}
          modalStatus={modalOpen}
          error={errorState}
        ></HintModal>
        {children}
      </>
    </SessionContext.Provider>
  );
}
