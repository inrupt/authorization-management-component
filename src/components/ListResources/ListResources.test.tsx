// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { act, render } from "@testing-library/react";
import * as axe from "axe-core";
import ListResources from "./ListResources";
import SessionProvider from "../../session/SessionProvider.mock";

describe("ListResources", () => {
  it("renders a ListResources component", async () => {
    const { asFragment, container } = render(
      <SessionProvider mockAccessEndpoint="https://example.org">
        <ListResources
          key="https://example.org/"
          resourceIri="https://example.org/"
          type="Container"
          name="Test Container"
          contains={async () => [
            {
              type: "Container",
              name: "Test Sub Container",
              contains: async () => [
                {
                  type: "Container",
                  name: "Test Nested Sub Container",
                  contains: async () => [],
                  key: "https://example.org/key3",
                  resourceIri: "https://example.org/key3",
                },
              ],
              key: "https://example.org/key",
              resourceIri: "https://example.org/key",
            },
            {
              type: "Resource",
              name: "Test Sub Resource",
              contains: async () => [],
              key: "https://example.org/key2",
              resourceIri: "https://example.org/key2",
            },
          ]}
        />
      </SessionProvider>,
    );
    await act(async () => {
      /* noop */
    });
    expect(asFragment()).toMatchSnapshot();

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });
});
