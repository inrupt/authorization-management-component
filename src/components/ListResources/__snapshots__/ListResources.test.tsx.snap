// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListResources renders a ListResources component 1`] = `
<DocumentFragment>
  <div
    class="resource-div"
  >
    <div
      class="adjacent-icon"
      style="margin-left: 0px; padding-right: 0px;"
    />
    <div
      class="name-content"
    >
      Test Container
    </div>
    <i
      class="bi bi-chevron-right chevron-icon"
      role="button"
      tabindex="0"
      title="expand"
    />
  </div>
</DocumentFragment>
`;
